functions/index.js

```diff
@@ -84,74 +69,67 @@ async function handleRequest(req, res, payloadHandler) {
 
   payload.token = token;
 
-  const ref = db.collection('rateLimits').doc(today).collection('tokens').doc(token);
-
-  let docExists = false;
-  let docData = {
-    attemptsCount: 0,
-    deliveredCount: 0,
-    errorCount: 0,
-    totalCount: 0,
-    expiresAt: getFirestoreTimestamp(),
-  };
-
+  // Create a rate limiter instance for this request
+  const rateLimiter = new RateLimiter(token, MAX_NOTIFICATIONS_PER_DAY, debug);
+  
+  let rateLimitInfo;
   try {
-    const currentDoc = await ref.get();
-    docExists = currentDoc.exists;
-    if (currentDoc.exists) {
-      docData = currentDoc.data();
-    }
+    rateLimitInfo = await rateLimiter.checkRateLimit();
   } catch (err) {
     return handleError(req, res, payload, 'getRateLimitDoc', err);
   }
 
-  docData.attemptsCount = docData.attemptsCount + 1;
-
-  if (updateRateLimits && docData.deliveredCount === MAX_NOTIFICATIONS_PER_DAY) {
-    try {
-      await sendRateLimitedNotification(token);
-    } catch (err) {
-      handleError(req, res, payload, 'sendRateLimitNotification', err, false);
+  if (updateRateLimits) {
+    // Increment attempts count
+    const attemptInfo = await rateLimiter.recordAttempt();
+    
+    if (attemptInfo.shouldSendRateLimitNotification) {
+      try {
+        await sendRateLimitedNotification(token);
+      } catch (err) {
+        handleError(req, res, payload, 'sendRateLimitNotification', err, false);
+      }
     }
-  }
 
-  if (updateRateLimits && docData.deliveredCount > MAX_NOTIFICATIONS_PER_DAY) {
-    await setRateLimitDoc(ref, docExists, docData, req, res);
-    return res.status(429).send({
-      errorType: 'RateLimited',
-      message: 'The given target has reached the maximum number of notifications allowed per day. Please try again later.',
-      target: token,
-      rateLimits: getRateLimitsObject(docData),
-    });
+    if (attemptInfo.isRateLimited) {
+      return res.status(429).send({
+        errorType: 'RateLimited',
+        message: 'The given target has reached the maximum number of notifications allowed per day. Please try again later.',
+        target: token,
+        rateLimits: attemptInfo.rateLimits,
+      });
+    }
   }
 
-  docData.totalCount = docData.totalCount + 1;
-
   if (debug) functions.logger.info('Sending notification', { notification: JSON.stringify(payload) });
```

Comment:
```suggestion
  if (debug) {
    functions.logger.info('Sending notification', { notification: JSON.stringify(payload) });
  }
```
Missed in the previous RP